File: src/services/dingtalk.py
"""
DingTalk service for message handling and API calls.
"""

import base64
import hashlib
import hmac
import os
import time
from typing import Any

import httpx
from dotenv import load_dotenv

load_dotenv(override=True)

DINGTALK_TOKEN_URL = "https://oapi.dingtalk.com/gettoken"
DINGTALK_SEND_URL = (
    "https://oapi.dingtalk.com/topapi/message/corpconversation/asyncsend_v2"
)


class DingTalkService:
    """DingTalk service for handling messages and API calls."""

    def __init__(
        self,
        app_key: str | None = None,
        app_secret: str | None = None,
        agent_id: str | None = None,
    ):
        # Use explicit values when provided (even empty string), fall back to env only when None
        self.app_key = app_key if app_key is not None else os.getenv("DINGTALK_APP_KEY")
        self.app_secret = app_secret if app_secret is not None else os.getenv("DINGTALK_APP_SECRET")
        self.agent_id = agent_id if agent_id is not None else os.getenv("DINGTALK_AGENT_ID")
        self._access_token: str | None = None
        self._token_expires_at: float = 0

    def verify_signature(self, timestamp: str, signature: str, nonce: str) -> bool:
        """Verify DingTalk message signature.

        Args:
            timestamp: Message timestamp
            signature: Message signature to verify
            nonce: Random nonce

        Returns:
            True if signature is valid
        """
        if not self.app_secret:
            return True  # Skip verification if no secret configured

        string_to_sign = f"{self.app_secret}{timestamp}{nonce}"
        hmac_code = hmac.new(
            string_to_sign.encode("utf-8"), digestmod=hashlib.sha256
        ).digest()
        my_signature = base64.b64encode(hmac_code).decode("utf-8")
        return my_signature == signature

    def get_access_token(self) -> str:
        """Get DingTalk access token, using cache when valid.

        Returns:
            Access token string

        Raises:
            Exception: If credentials not configured or API returns error
        """
        if not self.app_key or not self.app_secret:
            raise Exception("DingTalk credentials not configured")

        # Return cached token if still valid
        if self._access_token and time.time() < self._token_expires_at:
            return self._access_token

        response = httpx.get(
            DINGTALK_TOKEN_URL,
            params={"appkey": self.app_key, "appsecret": self.app_secret},
            timeout=10,
        )
        response.raise_for_status()
        data = response.json()

        if data.get("errcode") != 0:
            raise Exception(f"DingTalk token error: {data.get('errmsg', 'unknown')}")

        self._access_token = data["access_token"]
        expires_in = data.get("expires_in", 7200)
        self._token_expires_at = time.time() + expires_in - 300  # 5-minute buffer

        return self._access_token

    def send_message(self, user_id: str, msg_type: str, content: str) -> dict[str, Any]:
        """Send message to a DingTalk user via work notification.

        Args:
            user_id: Target user's staff ID
            msg_type: Message type — "text" or "markdown"
            content: Message content

        Returns:
            Dict with code=0 on success

        Raises:
            Exception: If API returns an error code
        """
        token = self.get_access_token()

        if msg_type == "markdown":
            lines = content.split("\n")
            title = lines[0].lstrip("#").strip() if lines else "消息"
            msg = {"msgtype": "markdown", "markdown": {"title": title, "text": content}}
        else:
            msg = {"msgtype": "text", "text": {"content": content}}

        payload = {
            "agent_id": self.agent_id,
            "userid_list": user_id,
            "msg": msg,
        }

        response = httpx.post(
            DINGTALK_SEND_URL,
            params={"access_token": token},
            json=payload,
            timeout=10,
        )
        response.raise_for_status()
        data = response.json()

        if data.get("errcode") != 0:
            raise Exception(f"DingTalk send error: {data.get('errmsg', 'unknown')}")

        return {"code": 0, "msg": "success", "user_id": user_id, "msg_type": msg_type}

    def parse_webhook_message(self, body: dict[str, Any]) -> dict[str, Any]:
        """Parse incoming webhook message.

        Args:
            body: Webhook request body

        Returns:
            Parsed message with type, content, user_id
        """
        msg_type = body.get("msgtype", "text")

        result = {
            "msg_type": msg_type,
            "user_id": body.get("senderStaffId", body.get("userId", "")),
            "content": "",
            "conversation_id": body.get("conversationId", ""),
        }

        if msg_type == "text":
            text_content = body.get("text", {})
            result["content"] = text_content.get("content", "")
        elif msg_type == "voice":
            voice_content = body.get("voice", {})
            result["media_id"] = voice_content.get("mediaId", "")
        elif msg_type == "markdown":
            markdown_content = body.get("markdown", {})
            result["title"] = markdown_content.get("title", "")
            result["content"] = markdown_content.get("text", "")

        return result


def get_dingtalk_service() -> DingTalkService:
    """Get singleton DingTalk service instance."""
    return DingTalkService()

File: src/services/dingtalk.ts
import * as crypto from 'crypto';
import { z } from 'zod';
import { config } from '../config.js';

// DingTalk webhook message schema
const DingTalkMessageSchema = z.object({
  msgtype: z.string(),
  text: z.object({ content: z.string() }).optional(),
  content: z.object({ text: z.string() }).optional(),
  voice: z.object({
    media_id: z.string(),
    duration: z.number().optional(),
    recognition: z.string().optional(),
  }).optional(),
  senderStaffId: z.string().optional(),
  senderId: z.string().optional(),
  conversationId: z.string().optional(),
  chatbotUserId: z.string().optional(),
  sessionWebhook: z.string().optional(),
  sessionWebhookExpiredTime: z.number().optional(),
});

export type DingTalkMessage = z.infer<typeof DingTalkMessageSchema>;

export interface ParsedMessage {
  msg_type: string;
  user_id: string;
  content: string;
  media_id?: string;
  conversation_id?: string;
}

// Callback message interface for Stream API (moved from ./dingtalk/types.ts)
export interface CallbackMessage {
  msgId: string;
  createAt: number;
  senderStaffId: string;
  conversationId: string;
  conversationType: '1' | '2' | '3';
  senderCorpId: string;
  conversationTitle: string;
  msgtype: string;
  text?: { content: string };
  image?: { media_id: string };
  voice?: { media_id: string; duration: number };
  file?: { media_id: string };
  location?: { title: string; address: string; latitude: number; longitude: number };
  link?: { title: string; text: string; message_url: string; pic_url?: string };
  markdown?: { title: string; text: string };
}

// Message types for sending
interface TextMessage {
  msgtype: 'text';
  text: { content: string };
}

interface MarkdownMessage {
  msgtype: 'markdown';
  markdown: { title: string; text: string };
}

interface LinkMessage {
  msgtype: 'link';
  link: { title: string; text: string; messageUrl: string; picUrl?: string };
}

type OutboundMessage = TextMessage | MarkdownMessage | LinkMessage;

interface TokenResponse {
  errcode: number;
  errmsg: string;
  access_token?: string;
  expires_in?: number;
}

interface SendMessageResponse {
  errcode: number;
  errmsg: string;
  task_id?: number;
}

export class DingTalkService {
  private readonly appKey: string;
  private readonly appSecret: string;
  private readonly agentId: string;
  private accessToken: string | null = null;
  private accessTokenExpireTime: number = 0;

  constructor(appKey: string, appSecret: string, agentId: string) {
    this.appKey = appKey;
    this.appSecret = appSecret;
    this.agentId = agentId;
  }

  /**
   * Verify DingTalk webhook signature
   */
  verifySignature(timestamp: string, signature: string, nonce: string): boolean {
    const signString = `${timestamp}\n${nonce}`;
    const hmac = crypto.createHmac('sha256', this.appSecret);
    const calculatedSignature = hmac.update(signString).digest('base64');
    return calculatedSignature === signature;
  }

  /**
   * Parse DingTalk webhook message
   */
  parseWebhookMessage(body: unknown): ParsedMessage {
    const parsed = DingTalkMessageSchema.parse(body);

    const msg_type = parsed.msgtype || 'text';
    const user_id = parsed.senderStaffId || parsed.senderId || '';
    let content = '';
    let media_id: string | undefined;

    if (msg_type === 'text') {
      content = parsed.text?.content || parsed.content?.text || '';
    } else if (msg_type === 'voice') {
      media_id = parsed.voice?.media_id;
      content = parsed.voice?.recognition || '';
    }

    return {
      msg_type,
      user_id,
      content,
      media_id,
      conversation_id: parsed.conversationId,
    };
  }

  /**
   * Get access token from DingTalk API with caching
   */
  private async getAccessToken(): Promise<string> {
    const now = Date.now();

    // Return cached token if still valid
    if (this.accessToken && now < this.accessTokenExpireTime) {
      return this.accessToken;
    }

    try {
      const response = await fetch(
        `https://oapi.dingtalk.com/gettoken?appkey=${this.appKey}&appsecret=${this.appSecret}`
      );

      const data = (await response.json()) as TokenResponse;

      if (data.errcode !== 0) {
        throw new Error(`Failed to get access token: ${data.errmsg}`);
      }

      this.accessToken = data.access_token || '';
      // Subtract 1 minute to be safe
      this.accessTokenExpireTime = now + ((data.expires_in || 7200) * 1000 - 60000);

      return this.accessToken;
    } catch (error) {
      console.error('Error getting access token:', error);
      throw error;
    }
  }

  /**
   * Send message via DingTalk API
   */
  private async sendMessage(
    params: {
      userid?: string;
      unionid?: string;
      openConversationId?: string;
      cid?: string;
    },
    msg: OutboundMessage
  ): Promise<SendMessageResponse> {
    const accessToken = await this.getAccessToken();

    const queryParams = new URLSearchParams({
      agent_id: this.agentId,
    });
    
    if (params.userid) queryParams.set('userid', params.userid);
    if (params.unionid) queryParams.set('unionid', params.unionid);
    if (params.openConversationId) queryParams.set('openConversationId', params.openConversationId);
    if (params.cid) queryParams.set('cid', params.cid);

    try {
      const response = await fetch(
        `https://oapi.dingtalk.com/topapi/message/corpconversation/asyncsend_v2?access_token=${accessToken}&${queryParams.toString()}`,
        {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({ msg }),
        }
      );

      const data = (await response.json()) as SendMessageResponse;

      if (data.errcode !== 0) {
        throw new Error(`Failed to send message: ${data.errmsg}`);
      }

      return data;
    } catch (error) {
      console.error('Error sending message:', error);
      throw error;
    }
  }

  /**
   * Send text message to DingTalk webhook
   */
  async sendTextMessage(webhookUrl: string, content: string): Promise<void> {
    try {
      const response = await fetch(webhookUrl, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          msgtype: 'text',
          text: { content },
        }),
      });

      if (!response.ok) {
        console.error('Failed to send DingTalk message:', response.status, response.statusText);
      }
    } catch (error) {
      console.error('Error sending DingTalk message:', error);
    }
  }

  /**
   * Send text message to a user
   */
  async sendTextToUser(userId: string, content: string): Promise<SendMessageResponse> {
    const msg: TextMessage = {
      msgtype: 'text',
      text: { content },
    };
    return this.sendMessage({ userid: userId }, msg);
  }

  /**
   * Send markdown message to a user
   */
  async sendMarkdownToUser(userId: string, title: string, text: string): Promise<SendMessageResponse> {
    const msg: MarkdownMessage = {
      msgtype: 'markdown',
      markdown: { title, text },
    };
    return this.sendMessage({ userid: userId }, msg);
  }

  /**
   * Send text message to a conversation
   */
  async sendTextToConversation(conversationId: string, content: string): Promise<SendMessageResponse> {
    const msg: TextMessage = {
      msgtype: 'text',
      text: { content },
    };
    return this.sendMessage({ cid: conversationId }, msg);
  }

  /**
   * Send markdown message to a conversation
   */
  async sendMarkdownToConversation(conversationId: string, title: string, text: string): Promise<SendMessageResponse> {
    const msg: MarkdownMessage = {
      msgtype: 'markdown',
      markdown: { title, text },
    };
    return this.sendMessage({ cid: conversationId }, msg);
  }

  /**
   * Send message to user (legacy interface - supports text/markdown message objects)
   */
  async sendToUser(userId: string, msg: { msgtype: string; text?: { content: string }; content?: string }): Promise<void> {
    if (msg.msgtype === 'text') {
      const content = msg.text?.content || msg.content || '';
      await this.sendTextToUser(userId, content);
    } else if (msg.msgtype === 'markdown') {
      const content = (msg as any).markdown?.text || msg.text?.content || msg.content || '';
      await this.sendMarkdownToUser(userId, content, content);
    }
  }

  /**
   * Send message to conversation (legacy interface - supports text/markdown message objects)
   */
  async sendToConversation(conversationId: string, msg: { msgtype: string; text?: { content: string }; content?: string }): Promise<void> {
    if (msg.msgtype === 'text') {
      const content = msg.text?.content || msg.content || '';
      await this.sendTextToConversation(conversationId, content);
    } else if (msg.msgtype === 'markdown') {
      const content = (msg as any).markdown?.text || msg.text?.content || msg.content || '';
      await this.sendMarkdownToConversation(conversationId, content, content);
    }
  }
}

// Singleton instance
let dingtalkServiceInstance: DingTalkService | null = null;

export function getDingtalkService(): DingTalkService {
  if (!dingtalkServiceInstance) {
    dingtalkServiceInstance = new DingTalkService(
      config.DINGTALK_APP_KEY,
      config.DINGTALK_APP_SECRET,
      config.DINGTALK_AGENT_ID,
    );
  }
  return dingtalkServiceInstance;
}
